pallets/maintenance-mode/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A pallet to put your runtime into a restricted maintenance or safe mode. This is useful when
+//! performing site maintenance, running data migrations, or protecting the chain during an attack.
+//!
+//! This introduces one storage read to fetch the base filter for each extrinsic. Hoever, it should
+//! be that the state cache eliminates this cost almost entirely. I wonder if that can or should be
+//! reflected in the weight calculation.
```

Comment:
That's an idea I've considered before (reflecting the fact that things are likely to be cached, or that multiple reads/writes to the same item will be cheaper).

Is this comment related to the `enter_maintenance_mode` and `resume_normal_operation` calls, or to the fact that this overhead will effectively be added to other calls? The latter is rather concerning, as it basically adds this cost to everything.

Reply:
Yeah, it's for every single extrinsic. We have to know whether the call passes the filter, and now that the filter depends whether we're in maintenance mode, we have to check that first, which means a storage read. I tried hard to find a way that we only incur the read when we're in maintenance mode, but I couldn't think of a way.

Reply:
I think your assumption about it being cached is pretty important, then. That might be worth explicitly testing at some point...

Reply:
So the db::read is done at each extrinsic right ?
In that case we should cache it at the block level (considering it is a possibility). We could make it an on_initialize to read it.

Reply:
(I mean for the next version)

Reply:
@crystalin You mean read it once in `on_initialize` so it is already in state cache? Is that better then letting it get cached in the first actual usage?

I guess maybe it is in terms of weight. We can account for one cold storage read worth of weight in `on_initialize` and then account only for a warm storage read worth of weight in each extrinsic.

Speaking of each extrinsic. I guess we need to re-benchmark them all in this configuration in order to account for this weight?